Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/1-basic-syntax/1-defining-an-actor.md

  var count : Nat = 0;

  // Define a publically callable function that when called 
  // will greet someone and increment the counter.
  public func greet(name : Text) : async Text {
    count += 1;
    "Hello, " # name # "! You are visitor number " # debug_show(count);
  };

  // Define a publically called function to 
  // return the current value of 'count' separately.
  public query func readCount() : async Nat {
    count
  };
};

// Make a call to the greet function with the input of "Programmer"
await Main.greet("Programmer");
```

A Motoko actor always presents its interface as a suite of named [functions](https://internetcomputer.org/docs/motoko/fundamentals/basic-syntax/functions) (also called methods) with defined argument and return types. When Motoko code is compiled, this interface is automatically translated to [Candid](https://internetcomputer.org/docs/building-apps/interact-with-canisters/candid/candid-concepts), an interface description language. The Candid description can be consumed by other canisters, including canisters written in another language such as Rust.

The above example's corresponding Candid interface can be found below.

```did
service : {

>   greet : (text) -> (text) query;

Reviewer: ```suggestion
  greet : (text) -> (text);
  readCount : () -> (nat) query;
```

# Defining an actor

> In Motoko, an `actor` is a unit of computation that encapsulates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Unlike traditional functions or objects in other programming languages, actors operate independently and communicate via [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) messaging. Each actor maintains its own message queue, enabling concurrent execution.

Reviewer 2: ```suggestion
In Motoko, an **actor** is a unit of computation that encapsulates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Unlike traditional functions or objects in other programming languages, actors operate independently and communicate via [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) messaging. Each actor maintains its own message queue, enabling concurrent execution.
```

In Motoko, actors are defined at the top level of a source file using the `actor` keyword. Public functions within an actor must be marked `shared` and return `async` types to support remote, asynchronous calls.

An actor definition is required for a Motoko program to be deployed as a canister on ICP.

```motoko
// Declares an actor named Main.
// The actor is defined as 'persistent,' meaning its state will persist. 
// Learn more about persistence: [link]
persistent actor Main {
  // Define a variable called 'count' to track the number of greetings.
  var count : Nat = 0;

  // Define a publically callable function that when called 
  // will greet someone and increment the counter.
  public func greet(name : Text) : async Text {
    count += 1;
    "Hello, " # name # "! You are visitor number " # debug_show(count);
  };

  // Define a publically called function to 
  // return the current value of 'count' separately.
  public query func readCount() : async Nat {
    count
  };
};

// Make a call to the greet function with the input of "Programmer"

> await Main.greet("Programmer");

Reviewer 2: Remove await so code compiles in ninja/dfx

> await Main.greet("Programmer");

Reviewer 2: ```suggestion
```

# Defining an actor

> In Motoko, an **actor**, declared with the `actor` keyword, is a computational process with its own  [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior.

Reviewer 2: ```suggestion
In Motoko, an **actor** is a computational process with its own [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Actors are declared with the `actor` keyword, 
```

> In Motoko, an **actor**, declared with the `actor` keyword, is a computational process with its own  [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior.

Reviewer 2: ```suggestion
In Motoko, an `actor` is a unit of computation that encapsulates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Unlike traditional functions or objects in other programming languages, actors operate independently and communicate via [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) messaging. Each actor maintains its own message queue, enabling concurrent execution.

You should define an actor when you want to encapsulate state and expose a public API that can be accessed asynchronously by other actors, canisters, or external clients.

More specifically, define an actor when:

- You are building a canister smart contract that maintains private state and exposes public functions.
- You want to create an application that runs on the Internet Computer and is accessible by users or other canisters.
- You want to take advantage of the actor model's benefits, such as memory isolation, single-threaded execution for update calls (avoiding race conditions), and asynchronous communication.

In Motoko, actors are defined at the top level of a source file using the `actor` keyword. Public functions within an actor must be marked `shared` and return `async` types to support remote, asynchronous calls.

An actor definition is required for a Motoko program to be deployed as a canister on ICP.
```

Reviewer 2: This was my previous suggestion for an edit but it appears to have been discarded. Please use this wording instead of the current text.

> The state is defined by the actor's private variables and the behavior is defined by its public functions - the functions it exposes to other actors.

Reviewer 2: ```suggestion
The state of an actor is determined by its private variables, while its behavior is defined by the public functions it exposes to other actors.
```

> The state is defined by the actor's private variables and the behavior is defined by its public functions - the functions it exposes to other actors.

Reviewer 2: ```suggestion

```

> 

Reviewer 2: In Motoko, an `actor` is a unit of computation that encapsulates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Unlike traditional functions or objects in other programming languages, actors operate independently and communicate via [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) messaging. Each actor maintains its own message queue, enabling concurrent execution.

You should define an actor when you want to encapsulate state and expose a public API that can be accessed asynchronously by other actors, canisters, or external clients.

More specifically, define an actor when:

- You are building a canister smart contract that maintains private state and exposes public functions.
- You want to create an application that runs on the Internet Computer and is accessible by users or other canisters.
- You want to take advantage of the actor model's benefits, such as memory isolation, single-threaded execution for update calls (avoiding race conditions), and asynchronous communication.

In Motoko, actors are defined at the top level of a source file using the `actor` keyword. Public functions within an actor must be marked `shared` and return `async` types to support remote, asynchronous calls.

An actor definition is required for a Motoko program to be deployed as a canister on ICP.

An actor operates independently but can communicate with other actors by calling their public functions. A call to a public function sends an [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) message to the receiving actor.

> Each actor maintains its own queues of incoming messages, one queue per sender. Message are processed in order, so that the processing of one message cannot interfere with the processing of another, protecting the actor's state from concurrent modification.
> 
> Because each actor processes its messages independently, multiple actors can execute messages at the same time, allowing for parallel execution at the level of actors.

Reviewer 2: ```suggestion
Each actor maintains separate queues of incoming messages, one per sender. Messages are processed in order, ensuring that one message cannot interfere with another. This protects the actor's state from concurrent modification.

Since actors process messages independently, multiple actors can handle messages in parallel, enabling concurrent execution across actors.
```

> Although a Motoko actor can have many internal components, including functions, classes and modules, the unit of deployment on ICP is always an actor. If you want to deploy Motoko code as an  ICP canister that code must define an actor.
> 
> Every Motoko actor corresponds to an ICP canister. Conversely, any ICP canister, regardless of the implementation language is represented as an actor when imported into Motoko.

Reviewer 2: ```suggestion

```

```motoko name=Main

> // Declares an actor named Main.

Reviewer 2: This example returns the output 
```
`ys6dh-5cjiq-5dc` :
  actor {
    greet : shared Text -> async Text;
    readCount : shared query () -> async Nat
  }
```

IMO this is very confusing to the reader as the first fundamental example shown to them. Can we make the output something different?

persistent actor Main {
  // Define a private variable called 'count' to track the number of greetings.
  var count : Nat = 0;

  // Define a public function that asynchronously returns a greeting
  // and increments the counter.
  public func greet(name : Text) : async Text {
    count += 1;
    "Hello, " # name # "! You are visitor number " # debug_show(count);
  };

  // Define a publically called function to 
  // return the current value of 'count' separately.
  public query func readCount() : async Nat {
    count
  };
};
```

:::info
This code defines an actor that can be deployed on ICP.
The actor is declared as `persistent` so that its state, `count`, will be preserved
when the actor is upgraded.
<!---TODO update persistence link?---->
Learn more about [persistence](https://internetcomputer.org/docs/motoko/icp-features/data-persistence).
:::

Another actor can call `Main.greet()` with an argument and await the result:

> ```motoko include=Main

Reviewer 2: ```suggestion
```motoko include=Main no-repl
```

Reviewer 2: Using the run button on the preview gives an error and does not compile this

In Motoko, an actor is a unit of computation that encapsulates [state](https://internetcomputer.org/docs/motoko/fundamentals/state) and behavior. Unlike traditional functions or objects in other programming languages, actors operate independently and interact with each other through [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) messaging.

Actors are objects with special system-level capabilities. An actor supports asynchronous operations and can persist state across calls. Each actor runs independently with its own message queue, allowing concurrent execution.

A Motoko program typically starts by defining an actor.

```motoko
// Declares an actor named Motoko
persistent actor Motoko {
  // Defines a hello function that returns the message "Hello, name !"
  // name is a argument passed into the function of type Text
  public query func greet(name : Text) : async Text {
    "Hello, " # name # "!";
  };
};

> await Motoko.greet("Progrmmer");

Reviewer: ```suggestion
await Motoko.greet("Programmer");
```

```

A Motoko actor always presents its interface as a suite of named [functions](https://internetcomputer.org/docs/motoko/fundamentals/basic-syntax/functions) (also called methods) with defined argument and return types. When Motoko code is compiled, this interface is automatically translated to [Candid](https://internetcomputer.org/docs/building-apps/interact-with-canisters/candid/candid-concepts), an interface description language. The Candid description can be consumed by other canisters, written in Motoko or another language such as Rust.

> The above example's corresponding Candid interface can be found below.

Reviewer: Can you add a space between `greet` and `:` below, ie:

```did
service : {
  greet : (text) -> (text) query;
}
```
